### ISSUE TITLE
Where to put custom evaluation?

### ISSUE THREAD

**Issue by fratajcz** [uma-pi1/kge#162]
Hi!

I played around with LibKGE for a bit now and I like it very much, thanks for making it available!

However, I need to extend it in a certain way so the performance metrics are comparable to my previous experiments. For starters I want to use it with TransE, however that will be extended to other KGE models later.
I have a list of entities X, a list of relations Y and a second list of entities Z. I need to get the score of every possible combination of entities and relations (x_i,y_j,z_k) for every possible i,j and k. It doesn't matter by which of the j relations x_i and z_k are connected, I just take the maximum score over the j relations. I figured out how to do this with the trained model:

```
scores_list = []
for i in range(relation_Y_embeddings.shape[0]):
    scores = model._scorer.score_emb(entities_X_embeddings, relation_Y_embeddings[i].unsqueeze(0), entities_Z_embeddings, combine= "sp_")
    scores_list.append(scores.reshape((num_entities_X,num_entities_Z)))

best_scores, ind = torch.stack(scores_list,dim=2).max(2)
```

best_scores is now in the form of an adjacency matrix between entities X and entities Y where a high value corresponds to a high probability of an existing edge, which i compare to an actual adjacency matrix between entities X and entities Y for train, validation and testing by computing the MRR. 

Is there any way in which i can use this during evaluation when the model is being trained? Maybe also as a custom loss function in addition to the other losses? The validation set that is used during training is basically identical to the validation adjacency matrix that i mentioned, but as an edgelist of course. I have checked the repo, but I am not sure where exactly to put this custom feature and which parts of the code have to be changed to make it work during training.

I hope I have made the issue clear enough, but if not, don't hesitate to ask. 

Thanks!

**Comment by rgemulla** [comment]
Not sure I understand fully what you want to do, but generally:

1. You may add you own evaluation class and register it [here](https://github.com/uma-pi1/kge/blob/0347d4a36104a0adf6415469f07b0b4e121d28f9/kge/job/eval.py#L35).

2. You can add metrics to an existing evaluation job. To do this, add a function that computes and adds your metric to the post epoch hook of the respective job. See [here](https://github.com/uma-pi1/kge/blob/0347d4a36104a0adf6415469f07b0b4e121d28f9/kge/job/job.py#L185) for details.

Does this help already?

**Comment by samuelbroscheit** [comment]
> 1. You may add you own evaluation class and register it [here](https://github.com/uma-pi1/kge/blob/0347d4a36104a0adf6415469f07b0b4e121d28f9/kge/job/eval.py#L35).

Could we make this configurable like we do the type -> class translation for models and embedders? Should be easy to do and then users don't have to change core files.

**Comment by rgemulla** [comment]
Yes, I like this. We could use a `class_name` key as for the models, it's a quick small change. The custom evaluation job could then be easily added under the `user` key in the configuration.

**Comment by rgemulla** [comment]
Addressed in PR #164.

@fratajcz Does this already resolve your question?

**Comment by rgemulla** [comment]
Merged PR #164; see the updated section on extensibility in the README file. I am closing this issue, please reopen if needed.

**Comment by fratajcz** [comment]
@rgemulla Sorry that it took me so long for responding, but I managed to include my custom evaluation in the way you initially proposed and it works pretty fine! Thanks a lot for the help!
